src/arduinotrigger.cpp

```diff
@@ -8,33 +8,134 @@
+    });
+    checkConnectionTimer->start(
+        CHECK_CONNECTED_DELAY_MSEC); // TODO Use QSerial connected Signal
+}
+
+ArduinoTriggerWorker::~ArduinoTriggerWorker() {
+    delete serialInterface;
+    delete serialPeer;
+}
+
+void ArduinoTriggerWorker::sendSetupSlot(int m_cmdDelay, int m_frameRate,
+                                         int m_frameLimit, bool resetCounter) {
+    SetupStruct setup;
+    setup.delay_us = m_cmdDelay;
+    setup.pulse_hz = m_frameRate;
+    setup.pulse_limit = m_frameLimit;
+    setup.flags = RESET_COUNTER ? resetCounter : 0;
+    serialPeer->sendSetup(&setup);
+}
+
+void ArduinoTriggerWorker::recvSlot() {
+    uint8_t receive_buffer[GET_COBS_PACKET_BUFFER_SIZE];
+    unsigned int receive_buffer_len = 0;
+    do {
+        receive_buffer_len = serialInterface->get_cobs_packet(
+            reinterpret_cast<char *>(receive_buffer),
+            GET_COBS_PACKET_BUFFER_SIZE, GET_COBS_PACKET_TIMEOUT);
+
+        if (receive_buffer_len < MIN_LENGTH_MESSAGE) {
+            return;
+        }
+
+        serialPeer->handleCobsMessage(receive_buffer, receive_buffer_len);
+    } while (receive_buffer_len >= MIN_LENGTH_MESSAGE);
+}
 
 ArduinoTrigger::ArduinoTrigger(const QString &deviceName)
     : TriggerInterface{arduinoTrigger} {
     createSettings();
-    serialInterface = new SerialInterface(deviceName);
-    QTimer *timer = new QTimer(this);
-    connect(timer, SIGNAL(timeout()), this, SLOT(intitialStatusSlot()));
-    timer->setSingleShot(true);
-    timer->start(100);
+
+    workerClass = new ArduinoTriggerWorker(deviceName);
+    workerClass->moveToThread(&workerThread);
+    workerThread.start();
+
+    connect(workerClass, &ArduinoTriggerWorker::statusUpdated, this,
+            &ArduinoTrigger::statusUpdated);
+
+    connect(workerClass, &ArduinoTriggerWorker::provideTriggerdata, this,
+            &ArduinoTrigger::provideTriggerdata);
+
+    connect(this, &ArduinoTrigger::sendSetupSignal, workerClass,
+            &ArduinoTriggerWorker::sendSetupSlot);
 }
 
-ArduinoTrigger::~ArduinoTrigger() {
-    delete serialInterface;
-    delete m_triggerSettings;
+void ArduinoTrigger::enable() {
+    emit sendSetupSignal(m_cmdDelay, m_frameRate, m_frameLimit, true);
 }
 
-void ArduinoTrigger::intitialStatusSlot() {
-    if (serialInterface->isConnected()) {
-        emit statusUpdated(Ready, "");
+void ArduinoTrigger::pause(bool state) {
+    if (state) {
+        // Pause
+        emit sendSetupSignal(0, 0, 0, false);
     } else {
-        emit statusUpdated(Error, "Connection Failed!");
+        // Continue
+        emit sendSetupSignal(m_cmdDelay, m_frameRate, m_frameLimit, false);
     }
 }
 
+void ArduinoTrigger::disable() { emit sendSetupSignal(0, 0, 0, false); }
+
+ArduinoTrigger::~ArduinoTrigger() {
+
+    qDebug() << "Trying to stop TriggerCOM";
+    workerThread.requestInterruption();
+    workerThread.quit();
+    workerThread.wait();
+
+    // TODO: call super class destructor
+
```

Comment:
Add 'delete workerClass;' here to make sure the serial port gets closed properly and reconnecting the trigger is possible.

Reply:
I must have missed that, thanks for adding it in 92e8238a619f040cc6c593e10ae3f6792deb691a